src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java

```diff
@@ -273,70 +287,54 @@ static <T extends RandomGenerator> T of(String name, Class<T> category)
      */
     static <T extends RandomGenerator> RandomGeneratorFactory<T> factoryOf(String name, Class<T> category)
             throws IllegalArgumentException {
-        Provider<? extends RandomGenerator> uncheckedProvider = findProvider(name, category);
-        return new RandomGeneratorFactory<>(uncheckedProvider);
+        Class<? extends RandomGenerator> rgClass = findClass(name, category);
+        return new RandomGeneratorFactory<>(rgClass);
     }
 
     /**
-     * Fetch the required constructors for class of random number algorithm.
-     *
-     * @param randomGeneratorClass class of random number algorithm (provider)
+     * Ensure all the required constructors are fetched.
      */
-    private void getConstructors(Class<? extends RandomGenerator> randomGeneratorClass) {
-        if (ctor == null) {
-            synchronized (provider) {
-                if (ctor == null) {
-                    PrivilegedExceptionAction<Constructor<?>[]> ctorAction = randomGeneratorClass::getConstructors;
-                    try {
-                        @SuppressWarnings("removal")
-                        Constructor<?>[] ctors = AccessController.doPrivileged(ctorAction);
-
-                        Constructor<T> tmpCtor = null;
-                        Constructor<T> tmpCtorLong = null;
-                        Constructor<T> tmpCtorBytes = null;
-
-
-                        for (Constructor<?> ctorGeneric : ctors) {
-                            @SuppressWarnings("unchecked")
-                            Constructor<T> ctorSpecific = (Constructor<T>) ctorGeneric;
-                            final Class<?>[] parameterTypes = ctorSpecific.getParameterTypes();
-
-                            if (parameterTypes.length == 0) {
-                                tmpCtor = ctorSpecific;
-                            } else if (parameterTypes.length == 1) {
-                                Class<?> argType = parameterTypes[0];
-
-                                if (argType == long.class) {
-                                    tmpCtorLong = ctorSpecific;
-                                } else if (argType == byte[].class) {
-                                    tmpCtorBytes = ctorSpecific;
-                                }
+    private void ensureConstructors() {
+        if (ctor == null) {  // volatile load
+            synchronized (rgClass) {
```

Comment:
Here too I think we should synchronize on `this` - we would want to allow multiple different instances of a `RandomGeneratorFactory` for the same `RandomGenerator` class type to be able to concurrently instantiate their individual instance fields (like the `ctor`(s) and `properties`).

Reply:
Then I would even remove the double-checking idiom, the `volatile` on `ctor` and `properties`, and declare methods `getProperties()` and `ensureConstructors()` as `synchronized`.
I'm not sure that the double-checking optimization brings much value on contemporary JVMs.

But I feel that the followup PR discussed before wouldn't need `synchronized` at all.

WDYT?

Reply:
> Then I would even remove the double-checking idiom, the volatile on ctor and properties, and declare methods getProperties() and ensureConstructors() as synchronized.
> I'm not sure that the double-checking optimization brings much value on contemporary JVMs.

Making the methods synchronized would bring in a penalty that there will always be a monitor entry at every call site, even after the `properites` and `ctor`(s) are initialized. Ideally, we should just do all of this intialization in the constructor of the `RandomGeneratorFactory`, the one which takes the `Class<>` type of the `RandomGenerator`. We can then make the `properties` and the `ctor`(s) all `final` and not have to worry about any synchronization or volatile semantics. You would of course have to rework the ensureConstructors to not throw an exception at that time.

> But I feel that the followup PR discussed before wouldn't need synchronized at all.

Correct. The more I think about it, I think cleaning up all this in this PR itself might make both reviewing and the implementation a bit more simpler. What's your thoughts?

Reply:
OK, will do all the work in this PR.